### DISCUSSION TITLE
Only about 6600 pixels visible/used out of 7500 pixels

### DISCUSSION THREAD

**Discussion by msspad** [sn4k3/UVtools#436] [Q&A] [answered]
1) When slicing in PrusaSlicer with 7500 x 3240 pixels (Phrozen Mini 8k), opening the sl1 in UVTools, converting it to ctb v4, the models get shifted in the 7500-pixel axis. When I zoom out any layer with the scroll wheel, hold the CTRL-key and read the highest x-position of a pixel, it's 6569 and not 7500 pixels. Any idea what is wrong?

2) Calling `Operation.HaveROI` always returns false in `ScriptInit()` but works fine in `ScriptExecute()`. Is there a way to access `HaveROI` in  `ScriptInit()`?

3) Running the same script always requires quite a few clicks to go through. Could UVTools by default load the previously loaded script when opening the scripting dialog? It would save me a few clicks...

**Comment by sn4k3** [comment, marked as answer]
1. I'm not seeing same behaviour, looks correct to me. Also why converting to CTBv4? It should be the encrypted format.
![UVtools_2022-03-15_22-47-03](https://user-images.githubusercontent.com/113281/158484785-f6b34606-703d-4047-9a59-0e3d68e73cef.png)

2. ROI is passed on Process/Execute, so not available before. Do you really need it on the Init?

3. 
![image](https://user-images.githubusercontent.com/113281/158484891-b0054896-383f-499f-8581-22254d277e69.png)

**Reply by msspad** [reply]
1) 
- I noticed now that the "Layer image zoom level" is limited to 0.2. When fully zoomed at 0.2, the build plate is only partially visible. Could you increase this limit? I didn't even think I could pan the image when fully zoomed.
- I meant encrypted.ctb and not ctbv4, my mistake.
- the main issue remains that sl1 moves in the 7500-pixel axis, see screenshot. It might be an issue with PrusaSlicer?
![image](https://user-images.githubusercontent.com/6191764/158537698-0cad67ec-cb33-4b8c-bd9d-d9ea9b8f2bf4.png)

2) `HaveROI` in `ScriptInit()` is low priority, no problem if it's a hassle to implement. `ScriptInit()` is the only place to give the user a feedback, e.g. how the script is going to change things (e.g. apply matt surface, see last screenshot). That's the reason why it's somewhat useful to have.

3) I could not make it work with Settings>Tools>Keep... What I meant is that when opening the Scripting window below, the previously loaded script should be loaded. So the second screenshot instead of first.
![image](https://user-images.githubusercontent.com/6191764/158533304-e2c3cf3e-b326-4761-b689-a4838b404e45.png)
![image](https://user-images.githubusercontent.com/6191764/158533512-97dadeaa-cd9e-4aa6-bbc0-6285fd105362.png)

**Reply by sn4k3** [reply]
1.
> I noticed now that the "Layer image zoom level" is limited to 0.2. When fully zoomed at 0.2, the build plate is only partially visible. Could you increase this limit? I didn't even think I could pan the image when fully zoomed.

I can add 0.1x and 64x, but if you click on Resolution button it will show you all the area.

> the main issue remains that sl1 moves in the 7500-pixel axis, see screenshot. It might be an issue with PrusaSlicer?

My test was with **PS 2.4.1** but notice on thumbnail you have free left space, using equal objects is not good, because we never know orientation, you can still use rectangles, but label them with an out extrusion (1, 2, 3, 4) that way you know which represents what.
You can zoom out and select all image region with ROI, see it total size OR you can save layer image, go to properties and see it size
But my guess is that the object is not slicing on the limit

2.
> HaveROI in ScriptInit() is low priority, no problem if it's a hassle to implement. ScriptInit() is the only place to give the user a feedback, e.g. how the script is going to change things (e.g. apply matt surface, see last screenshot). That's the reason why it's somewhat useful to have.

I will have a look into that

3.
> I could not make it work with Settings>Tools>Keep... What I meant is that when opening the Scripting window below, the previously loaded script should be loaded. So the second screenshot instead of first.

You need to execute the tool in order to it save and reload next time

**Reply by msspad** [reply]
1)
- I don't know what a left-click on the button with "Zoom" in it does, but after clicking it, it's exactly the behaviour I would have expected in the first place.
- When I open a sl1 file in PrusaSlicer 2.4.1 where 3 objects are in the corner of the build plate, PrusaSlicer says in "Size" (right, bottom) "132.1 x 69.1 x 2". 132 should be about 164 or so. It seems that PrusaSlicer does not slice correctly.

2) alright, but it's really not that important

3) mmm, I don't know hey it's not working:
- Menu Tools > Scripting > button "loads an script file" (typo in tooltip :-) > choose "myscript.cs" > click "Scripting" > save file
- then open other file > Menu Tools > Scripting > "myscirpt.cs" is not loaded in window

**Reply by sn4k3** [reply]
> button with "Zoom"

The button you want is this:
![UVtools_2022-03-16_17-24-29](https://user-images.githubusercontent.com/113281/158650633-183b6480-e450-4b08-969f-5a6fd1cb0f3f.png)

>  It seems that PrusaSlicer does not slice correctly.

Well UVtools never lie :P


> then open other file > Menu Tools > Scripting > "myscirpt.cs" is not loaded in window

I just tried, it works for me.
You can't close UVtools, also your execution function must `return true` otherwise failed operations won't be considered

**Reply by msspad** [reply]
tested it again, the zoom level is weird after clicking those buttons. Just increase the zoom limit to 0.1 (or the max given by the width or height of the image and screen) so whenever the user scrolls in the image to the end, he can be sure to see the entire image/build plate. That would be great!


> Well UVtools never lie :P

How could I dare to doubt :-)
Though that leaves PrusaSlicer hardly useable and I don't think they care to fix this bug...



> You can't close UVtools

That's exactly the issue. It should reload the previously loaded script after restart of UVTools.

**Reply by sn4k3** [reply]
> Though that leaves PrusaSlicer hardly useable and I don't think they care to fix this bug...

See if you have any offset on machine settings. But if you open the SL1 file back and object is now moved then its a bug with them.

> That's exactly the issue. It should reload the previously loaded script after restart of UVTools.

Ah, yes, those are only valid within session.
All i can do is try to enable profiles with Script tool, with that you can set an default profile to use once open the tool.

**Reply by msspad** [reply]
> See if you have any offset on machine settings. 

will search today again, but could not find anything so far

> All i can do is try to enable profiles with Script tool, with that you can set an default profile to use once open the tool.

With profile you mean choosing a specific script file that then always loads when Scripting window opens? that would be great

**Reply by msspad** [reply]
> It seems that PrusaSlicer does not slice correctly.

ok, I found the real culprit, it's me :-) 
I put the wrong number for Bed x in PrusaSlicer
![image](https://user-images.githubusercontent.com/6191764/158782581-f9249699-f35e-445c-886b-c1cf89a00124.png)

**Reply by sn4k3** [reply]
Better that way ;)
UVtools also have profiles for that machine which you can copy

**Reply by msspad** [reply]
all 3 (zooming, HaveROI and script profile) work great in 3.1.0! 
I really appreciate your work, thanks!
